### PR TITLE
Move the default schema cache to the home directory

### PR DESCRIPTION
**Pull request by brendandburns**
@jlowdermilk 

https://github.com/kubernetes/kubernetes/issues/14070

### CODE REVIEW

File: pkg/kubectl/cmd/util/factory.go [hunk 1]
```diff
@@ -307,11 +308,28 @@ type schemaClient interface {
 	Get() *client.Request
 }
 
+func substituteUser(dir string) (string, error) {
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/util/factory.go]: IMO, we should do the same thing that we do for kubeconfig, which currently is to just use [os.Getenv(HOME)](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/loader.go#L42). If the goal is to be platform agnostic (*cough*windows*cough*), then I'd recommend we put this logic in the loader and sharing it for both schema caching and kubeconfig.

Also, it wasn't clear to me from docs whether `user.Current()` will detect an overwritten `$HOME`. Probably safest to do `os.GetEnv(HOME)` and fallback to this if it isn't set, so we don't break existing behavior.

**Comment by jlowdermilk** [comment]
If we're going to cache under home directory, I'd recommend we use `~/.kube/schema` as the default path, and share/refactor the existing logic we use for kubeconfig in [clientcmd/loader.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/loader.go)

File: pkg/kubectl/cmd/util/factory.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/util/factory.go]: done.

**Comment by brendandburns** [comment]
@jlowdermilk comments addressed, please re-check.

Thanks!
--brendan

File: pkg/client/unversioned/clientcmd/loader.go [hunk 2]
```diff
@@ -36,11 +36,12 @@ import (
...
 	RecommendedConfigPathFlag   = "kubeconfig"
 	RecommendedConfigPathEnvVar = "KUBECONFIG"
-	RecommendedHomeFileName     = "/.kube/config"
+	RecommendedHomeDir          = ".kube"
+	RecommendedFileName         = "config"
 )
 
 var OldRecommendedHomeFile = path.Join(os.Getenv("HOME"), "/.kube/.kubeconfig")
-var RecommendedHomeFile = path.Join(os.Getenv("HOME"), RecommendedHomeFileName)
+var RecommendedHomeFile = path.Join(os.Getenv("HOME"), RecommendedHomeDir, RecommendedFileName)
```
Comment (jlowdermilk) [inline pkg/client/unversioned/clientcmd/loader.go]: I would suggest adding
```
var RecommendedSchemaFile = path.Join(os.Getenv("HOME"), RecommendedHomeDir, "schema")
```
and using the var below as the `--schema-cache-dir` default value and in the help text.

File: pkg/kubectl/cmd/util/helpers.go [hunk 3]
```diff
@@ -270,7 +270,7 @@ func GetFlagDuration(cmd *cobra.Command, flag string) time.Duration {
 
 func AddValidateFlags(cmd *cobra.Command) {
 	cmd.Flags().Bool("validate", true, "If true, use a schema to validate the input before sending it")
-	cmd.Flags().String("schema-cache-dir", "/tmp/kubectl.schema", "If non-empty, load/store cached API schemas in this directory, default is '/tmp/kubectl.schema'")
+	cmd.Flags().String("schema-cache-dir", "~/.kube/schema", "If non-empty, load/store cached API schemas in this directory, default is '~/kubectl.schema', where '~' is replaced with $HOME by kubectl")
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/util/helpers.go]: As mentioned above, recommend using a var initialized in `clientcmd` as the default value and in the help text so there's only one place to keep it updated. Case in point, the help text here doesn't match the default value.

File: pkg/client/unversioned/clientcmd/loader.go (hunk 2 above)
Comment (brendandburns) [inline pkg/client/unversioned/clientcmd/loader.go]: done.

File: pkg/kubectl/cmd/util/helpers.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/util/helpers.go]: done.

**Comment by brendandburns** [comment]
Ok, I think this is ready to go.  @jlowdermilk ptal.

thanks!
--brendan

**Comment by jlowdermilk** [comment]
lgtm, thanks

**Comment by alex-mohr** [comment]
Jenkins:
15:01:31 	|  Ginkgo timed out waiting for all parallel nodes to report back!  |

Shippable:
error validating "docs/admin/limitrange/valid-pod.yaml": error validating data: link ~/.kube/schema/v1.1.0-alpha.1.935+72a8eb2feb2c3d/schema546582641 /tmp/tmp.EkmLhZKMC2/.kube/schema/v1.1.0-alpha.1.935+72a8eb2feb2c3d/api/v1/schema.json: no such file or directory; if you choose to ignore these errors, turn validation off with --validate=false
!!! Error in ./hack/test-cmd.sh:229
  'kubectl create "${kube_flags[@]}" -f docs/admin/limitrange/valid-pod.yaml' exited with status 1

**Comment by smarterclayton** [comment]
If the schema dir is not writable, does everything fail?  Or do we just cache in memory?  Home is possibly less commonly writable inside a container

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by jlowdermilk** [comment]
> If the schema dir is not writable, does everything fail? Or do we just cache in memory? Home is possibly less commonly writable inside a container

That's a good point. If running kubectl from container, should explicitly specify schema-dir. Actually, to handle that case better, it probably makes sense to allow you to specify schema dir via an env variable.

**Comment by smarterclayton** [comment]
We should probably tolerate not writable and fall back to in memory as
well.  If this is a read-only container....

On Fri, Sep 18, 2015 at 12:54 PM, Jeff Lowdermilk <notifications@github.com>
wrote:

> If the schema dir is not writable, does everything fail? Or do we just
> cache in memory? Home is possibly less commonly writable inside a container
>
> That's a good point. If running kubectl from container, should explicitly
> specify schema-dir. Actually, to handle that case better, it probably makes
> sense to allow you to specify schema dir via an env variable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14087#issuecomment-141506283>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by brendandburns** [comment]
Yeah, I agree we should prob. detect and tolerate writeable failure

File: pkg/kubectl/cmd/util/factory.go [hunk 4]
```diff
@@ -308,9 +308,16 @@ type schemaClient interface {
 	Get() *client.Request
 }
 
+func substituteUserHome(dir string) string {
+	if len(dir) == 0 || dir[0] != '~' {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory.go]: probably want to do a path split and check of the first segment to avoid replacing in strings like `~mydir`

**Comment by bgrant0607** [comment]
Test failures look real. Removing lgtm.

File: pkg/kubectl/cmd/util/factory.go (hunk 4 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/util/factory.go]: done.

**Comment by brendandburns** [comment]
Tests are sorted out @bgrant0607 @jlowdermilk I can haz lgtmz?

kthxbye

**Comment by bgrant0607** [comment]
I know that you're probably trying to reduce the number of review cycles, but it's impossible to tell what you changed if you squash prior to lgtm.

Anyway, reapplied lgtm since tests are passing and the PR previously had a lgtm from @jlowdermilk.

File: pkg/kubectl/cmd/util/factory_test.go [hunk 5]
```diff
@@ -302,3 +303,32 @@ func TestValidateCachesSchema(t *testing.T) {
 		t.Errorf("unexpected cache file error: %v", err)
 	}
 }
+
+func TestSubstitueUser(t *testing.T) {
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/util/factory_test.go]: Nit: TestSubstituteUser

**Pull request merged**